src/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs

```diff
@@ -238,15 +239,15 @@ public bool UseDefaultCredentials
                     throw new InvalidOperationException(SR.SmtpInvalidOperationDuringSend);
                 }
 
-                _transport.Credentials = value ? CredentialCache.DefaultNetworkCredentials : null;
+                _useDefaultCredentials = value;
```

Comment:
It looks like setting the `UseDefaultCredentials` property only impacts the credentials surfaced in SmtpClient but have no bearing on the underlying SmtpTransport state. Is this intentional?

Reply:
I didn't actually realize that, no it should probably set the transport state too since that is not being set anywhere, I'll take a look at that, how would you propose I add a test for that though?

Reply:
I'd just like to verify what behaviour we want here then:

1. If we set `Credentials`, we want the `_transport.Credentials` to be set to the`DefaultNetworkCredentials` if `UseDefaultCredentials == true`, otherwise we want it to be the newly set `Credentials` 
2. When setting `Credentials`, we only want the `_transpport.Credentials` to be set to the `Credentials` if we have `UseDefaultCredentials == false`

So with all that considered, when we use the `get` for `Credentials` do we want to return the Credentials we set or the current `_transport.Credentials`?

Reply:
I think it would be best to first clarify the expected behavior from an API point of view rather than how that translates to code changes for the PR.

The SmptClient API has, unfortunately, 2 separate properties for specifying the credentials.  it has the 'UseDefaultCredentials' bool property and a separate 'Credentials' property.  In hindsight, it would have been better to have a single property.

That said, how should the properties interact with each other? That really is the topic of the original issue and what this PR is attempting to address.

The simple explanation for expected behavior that I think of is:

* Each property should be able to be 'get' and 'set' in any order and the last 'set' value should always be retrievable accurately regardless of the order in which these two properties are used.

* 'UseDefaultCredentials' has priority over the 'Credentials' property. If 'UseDefaultCredentials' is true, then default credentials are used regardless of what is set in the 'Credentials' property.  If 'UseDefaultCredentials' is false, then the 'Credentitals' property is used.

Reply:
>how would you propose I add a test for that though?

Testing this with just unit tests is tricky. It really needs an end-to-end test. And that requires using an SMTP server that understand NTLM/Negotiate.  Negotiate is very hard because it requires the client and server to be joined on the same "domain" / "realm". We currently don't have a way to test that with our current CI (although work is in progress for expanding our Enterprise testing capabilities). NTLM is a little easier since the client doesn't have to be joined to a domain, but still requires some NTLM based SmtpServer.

We currently have a test SMTP server that is used in some of the unit tests. It's possible this could be expanded a little bit to at least test that the right NTLM challenge token was sent by SmtpClient.

But bottom line, we need some kind of "end to end" test here to validate that SmtpClient and related objects (like `_transport`) are correctly updated for the PR change.

Reply:
Okay great, so we're aligned on the expected behavior. 

I've been looking into testing this using a custom SMTP server but it does seem overkill to try to string one together to test that the credentials are being retained.

Can we not use reflection to access the `_transport.Credentials` data and compare that to what we'd set the `SmtpClient` with?

Reply:
>Can we not use reflection to access the _transport.Credentials data and compare that to what we'd set the SmtpClient with?

We normally don't like to use reflection since it complicates the tests many times and relies on internal implementation details that could change in the future (thus breaking the test).

But given the low likelihood of this architecture changing and the difficulty is testing this, I'm ok with using reflection for creating a test to verify this change.